src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala

```diff
@@ -52,10 +58,56 @@ class RegularSync(
       askForHeaders()
   }
 
-  def running: Receive = handleCommonMessages orElse {
-    case ResumeRegularSync =>
-      askForHeaders()
+  private def resumeRegularSync(): Unit = {
+    resumeRegularSyncTimeout.foreach(_.cancel)
+    resumeRegularSyncTimeout = None
+    askForHeaders()
+  }
 
+  /**
+    * Handles broadcasted blocks, processing them and inserting them to the blockchain if necessary
+    * FIXME: Current implementation is naive in that NewBlock msgs that can't be immediately processed are discarded
+    *        and requested again to the peers
+    */
+  def handleBroadcastedBlockMessages: Receive = {
+    case MessageFromPeer(NewBlock(newBlock, _), peerId) =>
+      //we allow inclusion of new block only if we are not syncing / reorganising chain
+      if (headersQueue.isEmpty && waitingForActor.isEmpty) {
+        val currentBestBlock = appStateStorage.getBestBlockNumber()
+        val maybeBlockParentTd = blockchain.getBlockHeaderByHash(newBlock.header.parentHash)
+          .flatMap(b => blockchain.getTotalDifficultyByHash(b.hash))
+        val maybeCurrentBlockTd = blockchain.getTotalDifficultyByNumber(currentBestBlock)
+
+        (maybeBlockParentTd, maybeCurrentBlockTd) match {
+          case (_, None) =>
+            //Something went wrong, we don't have the total difficulty of the latest block
+            //TODO: Investigate if we can recover from this error [EC-165]
```

Comment:
We have exactly same TODO error in line `302`, is this really the same case ? As i understand here we don't have the difficulty of currently best block saved in storage (is this even really possible?), and in line `302` we don't have the difficulty of the parent of some incoming block in the queue.

Reply:
> We have exactly same TODO error in line 302, is this really the same case ?

You are right, it's not exactly the same case, however it might be the case that both can be handled in a similar way, [as done in FastSync](https://github.com/input-output-hk/mantis/blob/phase/daedalus/src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala#L335). I'll add a comment on task EC-165 to split it into two if needed.

> is this even really possible?

AFAIK it shouldn't be, it should only be caused by data corruption or a bug on our client.